src/main/java/spoon/support/util/KeyedModelSet.java

```diff
@@ -7,135 +7,136 @@
-		set = Collections.synchronizedSet(new TreeSet<>(comparator));
+	protected KeyedModelSet() {
+		this.map = new ConcurrentSkipListMap<>();
 	}
 
 	protected abstract CtElement getOwner();
 	protected abstract CtRole getRole();
 	protected void onSizeChanged(int newSize) {
 	}
 
-	@Override
 	public int size() {
-		return set.size();
+		return map.size();
 	}
 
-	@Override
 	public boolean isEmpty() {
-		return set.isEmpty();
+		return map.isEmpty();
 	}
 
-	@Override
-	public boolean contains(Object o) {
-		return set.contains(o);
+	public boolean contains(K key) {
+		return map.containsKey(key);
 	}
 
-	@Override
-	public Object[] toArray() {
-		return set.toArray();
+	public Collection<T> values() {
+		return map.values();
 	}
 
-	@Override
-	public <T> T[] toArray(T[] a) {
-		return set.toArray(a);
+	public T get(K key) {
+		return map.get(key);
 	}
 
-	@Override
-	public boolean add(T e) {
+	public boolean add(K key, T e) {
 		if (e == null) {
 			return false;
 		}
 		CtElement owner = getOwner();
 		linkToParent(owner, e);
-		getModelChangeListener().onSetAdd(owner, getRole(), set, e);
+		getModelChangeListener().onSetAdd(owner, getRole(), new HashSet<>(map.values()), e);
 
 		// we make sure that the element is always the last put in the set
 		// for being least suprising for client code
-		if (set.contains(e)) {
-			set.remove(e);
-		}
+		map.put(key, e);
 
-		set.add(e);
 		return true;
 	}
 
-	@Override
-	public boolean remove(Object o) {
-		if (set.contains(o) == false) {
-			return false;
-		}
-		@SuppressWarnings("unchecked")
-		T e = (T) o;
-		getModelChangeListener().onSetDelete(getOwner(), getRole(), set, e);
-		if (set.remove(o) == false) {
-			throw new SpoonException("Element was contained in the Set, but Set#remove returned false. Not removed??");
+	public T remove(K key) {
+		T removed = map.remove(key);
+
+		if (removed == null) {
+			return null;
 		}
-		return true;
-	}
 
-	@Override
-	public boolean containsAll(Collection<?> c) {
-		return set.containsAll(c);
+		getModelChangeListener().onSetDelete(
+				getOwner(),
+				getRole(),
+				new HashSet<>(map.values()),
+				removed
+		);
+
+		return removed;
 	}
 
-	@Override
 	public void clear() {
-		if (set.isEmpty()) {
+		if (map.isEmpty()) {
 			return;
 		}
-		getModelChangeListener().onSetDeleteAll(getOwner(), getRole(), set, new LinkedHashSet<>(set));
-		set.clear();
+		getModelChangeListener().onSetDeleteAll(
+				getOwner(),
+				getRole(),
+				new HashSet<>(map.values()),
+				new LinkedHashSet<>(map.values())
+		);
+		map.clear();
 	}
 
 	@Override
 	public boolean equals(Object o) {
-		return set.equals(o);
+		return map.equals(o);
 	}
 
 	@Override
 	public int hashCode() {
-		return set.hashCode();
+		return map.hashCode();
 	}
 
-	@Override
 	public Iterator<T> iterator() {
 		return new Itr();
 	}
 
+	public void updateKey(K oldName, K newName) {
+		T type = map.remove(oldName);
+		if (type != null) {
```

Comment:
Updating a key that doesn't exist seems like an error to me.

Reply:
We are in a bit of a pickle here, as many elements are added *without a name*: The parent is set before the simple name while constructing the model...